### PR TITLE
large_enum_variants lint suggests to box variants above a configurable limit

### LINKED ISSUE #1127: Suggest to store large enum variants as Box
For example:

``` rust
enum Test {
    A(i32),
    B([i32; 8000]),
}
```

In this case it probably would be better to use `B(Box<[i32; 8000]>)`.

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #1127
obsoletes #1187

### CODE REVIEW

File: clippy_lints/src/large_enum_variant.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+//! lint when there are large variants on an enum
+
+use rustc::lint::*;
+use rustc::hir::*;
```
Comment (mcarton) [inline clippy_lints/src/large_enum_variant.rs]: Well, fortunately [you did not promise](https://github.com/Manishearth/rust-clippy/pull/1246#discussion_r81449854) :smile:

File: clippy_lints/src/large_enum_variant.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+use utils::span_help_and_lint;
+use rustc::ty::layout::TargetDataLayout;
+use rustc::ty::TypeFoldable;
+use rustc::traits::Reveal;
+
+/// **What it does:** Checks for large variants on enums.
```
Comment (mcarton) [inline clippy_lints/src/large_enum_variant.rs]: I prefer “enumeration” or “`enum`s”.

File: clippy_lints/src/large_enum_variant.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                                    .bytes()
+                            }
+                        })
+                        .sum();
+                    if size > self.maximum_variant_size_allowed {
+                        span_help_and_lint(cx,
+                                           LARGE_ENUM_VARIANT,
+                                           def.variants[i].span,
+                                           &format!("large enum variant found on variant `{}`", variant.name),
+                                           "consider boxing the large branches to reduce the total size of the enum");
```
Comment (mcarton) [inline clippy_lints/src/large_enum_variant.rs]: * Make a suggestion if there is only one field?
* s/branch/field

File: tests/compile-fail/large_enum_variant.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    C([T; 8000]),
+    D(T, [i32; 8000]), //~ ERROR large enum variant found on variant `D`
+}
+
+trait SomeTrait {
+    type Item;
+}
+
+enum LargeEnumGeneric<A: SomeTrait> {
+    Var(A::Item), // regression test, this used to ICE
```
Comment (mcarton) [inline tests/compile-fail/large_enum_variant.rs]: You add regression tests even before the PR is merged? :smile:

File: clippy_lints/src/large_enum_variant.rs (hunk 1 above)
Comment (oli-obk) [inline clippy_lints/src/large_enum_variant.rs]: in my defense, I only rebased this code ;)

File: tests/compile-fail/large_enum_variant.rs (hunk 4 above)
Comment (oli-obk) [inline tests/compile-fail/large_enum_variant.rs]: I got tired of fixing multiple commits during rebasing and simply merged them into one

File: clippy_lints/src/large_enum_variant.rs (hunk 1 above)
Comment (mcarton) [inline clippy_lints/src/large_enum_variant.rs]: In my defense, this does not appear in either the commit info (“@foo commented with @bar”), the first commit message, the blame info (you don't seem to have used `git rebase`?), or the PR's comment (“obsoletes #xyz” implies this is more than just a rebase).

**Comment by Arnavion** [comment]
Needs to be documented at https://github.com/Manishearth/rust-clippy/wiki#large_enum_variant

**Comment by Arnavion** [comment]
Also, this lint warns on all variants that exceed 200 bytes in size, as opposed to if the difference between the smallest and largest variant exceeding 200 bytes. Is that intentional? The latter seems to make more sense to me.

**Comment by Manishearth** [comment]
Updated the wiki.

**Pull request merged**